orders-accounting/src/price_calculation.rs

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) 2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use common::{
+    chain::{output_value::OutputValue, OrderId},
+    primitives::Amount,
+};
+use utils::ensure;
+
+use crate::{error::Result, Error, OrdersAccountingView};
+
+pub fn calculate_fill_order(
+    view: &impl OrdersAccountingView,
+    order_id: OrderId,
+    fill_value: &OutputValue,
+) -> Result<Amount> {
+    let order_data = view
+        .get_order_data(&order_id)
+        .map_err(|_| crate::Error::ViewFail)?
+        .ok_or(Error::OrderDataNotFound(order_id))?;
+    let ask_balance = view
+        .get_ask_balance(&order_id)
+        .map_err(|_| crate::Error::ViewFail)?
+        .ok_or(Error::OrderAskBalanceNotFound(order_id))?;
+    let give_balance = view
+        .get_give_balance(&order_id)
+        .map_err(|_| crate::Error::ViewFail)?
+        .ok_or(Error::OrderGiveBalanceNotFound(order_id))?;
+
+    {
+        let ask_balance = match order_data.ask() {
+            OutputValue::Coin(_) => OutputValue::Coin(ask_balance),
+            OutputValue::TokenV0(_) => unreachable!(),
```

Comment:
Please don't write code that crashes for branches that are disabled by forks and settings. Just return an error. This is spread everywhere.